docs/concepts/native-interop.rst

```diff
@@ -0,0 +1,254 @@
+  callback from unmanaged code. Notice how the LPARAM and HWND types are 
+  represented using ``IntPtr`` in the managed code. 
+* Lines #10 and #11 introduce the ``EnumWindows`` function from the user32.dll 
+  library. 
+* Lines #13 - 16 implement the delegate. For this simple example, we just want 
+  to output the handle to the console.
+* Finally, in line #19 we invoke the external method and pass in the delegate.
+
+Both of the above examples depend on parameters, and in both cases, the parameters 
+are given as managed types. Runtime does the "right thing" and processes these 
+into its equivalents on the other side. Since this process is really important 
+to writing quality native interop code, let's take a look at what happens when 
+the runtime *marshals* the types. 
+
+Type marshalling
+^^^^^^^^^^^^^^^^
+**Marshalling** is the process of packing up types when they need to cross the 
+managed boundary into native and vice versa. 
+
+Why is this needed? Simply put, the types in the managed and unmanaged world 
+are different. In managed world, for instance, you have a ``String``, while in 
+the unmanaged world strings can be Unicode ("wide"), non-Unicode, null-terminated, 
+ASCII, etc. By default, the .NET runtime will try to do the Right Thing and for
+many applications leaving it to its own devices is usually fine. 
+
+However, for those situations where you need extra control, you can employ the 
+``MarshalAs`` attribute to tell the runtime what is the expected type in the 
+unmanaged world. For instance, if we want the string to be sent as a null-terminated 
+ANSI string, we could do it like this:
+
+.. code-block:: c#
+    
+    [DllImport("somenativelibrary.dll"]
+    static extern int MethodA([MarshalAs(UnmanagedType.LPStr) string parameter);
+
+Another aspect of type marshalling is how to pass in a struct to an unmanaged method.
+For instance, some of the unmanaged methods require a struct as a parameter. 
+In these cases, we need to create a corresponding struct or a class in managed 
+part of the world to use it as a parameter. However, just defining the class is 
+not enough, we also need to instruct the marshaler how to map fields in the class 
+to the unmanaged struct. This is where the ``StructLayout`` attribute comes into 
+play. 
+
+.. code-block:: c#
+    :linenos:
+
+    [DllImport("kernel32.dll")]
+    static extern void GetSystemTime(SystemTime systemTime);
+
+    [StructLayout(LayoutKind.Sequential)]
+    class SystemTime {
+        public ushort Year;
+        public ushort Month;
+        public ushort DayOfWeek;
+        public ushort Day;
+        public ushort Hour;
+        public ushort Minute;
+        public ushort Second;
+        public ushort Milsecond;
+    }
+
+    public static void Main(string[] args) {
+        SystemTime st = new SystemTime();
+        GetSystemTime(st);
+        Console.WriteLine(st.Year);
+    }
+
+The example above shows off a simple example of calling into ``GetSystemTime()`` 
```

Comment:
Again, annotation of the sample would be better.

What's "pack-size boundaries"?